### DISCUSSION TITLE
Feature Request - Add Web Front End

### DISCUSSION THREAD

**Discussion by freedbygrace** [9tigerio/db2rest#694] [Ideas]
Add a web based front end that allows for the creation of connections, users, groups, jwt tokens, etc

Allows for querying of backend databases

Metrics to see how many requests from where etc

**Comment by thadguidry** [comment]
@freedbygrace Thanks for the ideas!  We have indeed been thinking about all of these already, and have plans to tackle them in various ways.
But I have some questions to understand your thinking and needs.

1. For adding connections, you mean, to have a webpage form that allows you to edit the environment variables for all the configuration parameters (DB_URL, DB_USER, etc...) ?
2. For querying, would you prefer something like [HTTPie for Web](https://httpie.io/app) / [HTTPie.AI](https://httpie.io/ai) ?  Where you not only could type and edit parameters and modify headers, but also with HTTPie.AI you could query with natural language, such as "Fetch all films starring Tom Cruise before the year 2000".
    - or do you envision some other design for querying?  If not my example above, what would your preferred design look like?  Can you point to examples you like or would prefer?
3. For metrics, I think it's best if we offer a robust API backend that users could use web hooks like from Splunk, Prometheus, Sensu, OpenNMS, Datadog, etc.  We could certainly make the packaging of this seamless as a choice during installation, where the choice would run a specific installer/config script for some of those, with initial default settings.  But I think this falls into very opinionated area, and so we'd likely only choose 1 as a default add-in script for our package release.  The community could submit add-in scripts that download, install, config them against our proposed API backend for DB2Rest metrics.
    - Caveats:  Networking issues, depending on where and how DB2Rest is hosted, so we have no control.  Again, here to, the community could work on scripts which we could host in another sister repo for managing DB2Rest add-in scripts that config and solve various networking/host issues.
4. Would you be willing to sponsor or pay for these features?
5. Would you be OK and pay for some of these features if offered only in an Enterprise version?

**Comment by kdhrubo** [comment]
Metrics and health already supported

https://db2rest.com/docs/category/actuator

**Comment by freedbygrace** [comment]
Hey! Thank you for the excellent feedback! I will answer questions in the same order.

1. Similar to how Dreamfactory works where there is a WebUI for database connection configuration to avoid having to mess around with configuration files etc. This is not necessarily bad and is quite useful for automation scenarios that do not require user interaction, but from a ease of consumption perspective, a WebUI would be great.
2. The HTTPie ideas actually look good and could actually be a better approach to what I had in mind, however, what I had originally thought of is a web based SQL query editor with intellisense, the ability to save queries, and export data to various formats.
3. Prometheus would work just fine! If the data can be exposed over http or https then we can slap that behind a reverse proxy.
4. I cannot sponsor, but I am definitely willing to make a donation.
5. Yes, I would not be opposed to this as I know this type of programming is more than a heavy lift!

**Comment by thadguidry** [comment]
@freedbygrace Thanks for answering those.  It helps us make informed decisions about the future direction of the project.

Regarding point 2. we actually try to avoid SQL somewhat since not all web developers are well versed in it, but they are quite comfortable when dealing with query URLs and parameters (hence the RQL syntax).  Our main users that we target are web developers, but knowing that infrastructure backend devs and support personnel, such as likely yourself, also need easy install and config options.

**Comment by kdhrubo** [comment]
Would move this to discussion.

**Comment by kdhrubo** [comment]
This is the most likely option to sustain the core project. 

1. Start a web control center project. (CC)
2. No UI component within the core DB2Rest project. 
3. CC can be used to control, manage, and query all the db2rest instances in your network and possibly in container management platform as well. 
4. CC will not be open source, neither free. This helps us compensate some of our key contributors who help in keeping DB2REST updated to latest dependencies, help make regular releases and keep adding various DB support.

**Comment by kdhrubo** [comment]
I think a simple UI with Vaadin may be good ?. @freedbygrace 
This way it can be used as standalone headless CMS.

**Reply by freedbygrace** [reply]
Something like that would be fantastic! Again, thank you all for even releasing and maintaining what you have done so far.

**Reply by thadguidry** [reply]
@freedbygrace Curious, what's the $ dollar value that you or others might put on something like we have been discussing, since you previously further above were not opposed to paying for this feature.

**Reply by freedbygrace** [reply]
It depends on your model I would say. Some sort of self hosted free tier would be great, but thereafter, a cloud provided solution that would generate passive income. For that, 2 to 5 per month, possibly more depth, for user, data sources, etc. If without cloud solution, a one time fee without ongoing support is another approach or with support for passive income. For me, $25 to $100 would be good, for attractiveness, but this is not to undervalue the product or work. Just throwing out numbers without full consideration of all the details.

**Reply by thadguidry** [reply]
Thanks!  Ok, just to clarify, for you, you would be comfortable paying a one-time fee of $25 to $100 for the feature of:
> a web based front end that allows for the creation of connections, users, groups, jwt tokens, etc
> Allows for querying of backend databases
> Metrics to see how many requests from where etc

And sorry, I didn't quite understand the "2 to 5 per month".  Is that $ dollars you would pay for the SaaS service with the above quoted features?

We care about you.  Don't worry about us or any models. :-)  It's highly useful to hear truthful feedback from you on what you find valuable, what feels too expensive, etc.

**Reply by thadguidry** [reply]
Oh, what if this proposed web front end was a full service, that would not only let you manage your own DB, but also come with an instance of a DB of your choice (which one do you like?) running in the cloud, and with DB2Rest connected to use that cloud DB, and also have the web frontend control with those features you asked for?  What's the value on that do you feel?

**Reply by freedbygrace** [reply]
Yes, perfect! From that perspective, I would say 2 to 5 dollars per month or those one time numbers because then I can spin up small projects without dealing with multiple solutions more complexity etc. In any case, in a company situation, I would naturally want to pay more because of the value it brings the company, especially if we come to rely upon it. For personal use, the same value is there but not at the same level because it's just me and family, maybe a couple friends using it, in my situations.

**Reply by freedbygrace** [reply]
To the second post, that is an amazing idea! But might get bad reception with people who like control over their backend and privacy concerns etc. That could easily be $50 to upwards of $100 per month though in my view.
